File: internal/application/userService.go
package application

import (
	"fmt"
	"visitor-management/internal/domain"
	"visitor-management/internal/ports"

	"github.com/google/uuid"
)

type UserService struct {
	repo ports.UserRepository
}

func NewUserService(repo ports.UserRepository) *UserService {
	return &UserService{repo: repo}
}

func (s *UserService) GetallUsers() ([]domain.User, error) {
	fmt.Println("UserService: calling repo.GetallUsers")
	users, err := s.repo.GetallUsers()
	if err != nil {
		return nil, err
	}

	var owners []domain.User
	for _, u := range users {
		if u.Role == "owner" {
			owners = append(owners, u)
		}
	}

	return owners, nil
}

func (s *UserService) GetUserByEmail(email string) (*domain.User, error) {
	return s.repo.GetByEmail(email)
}

func (s *UserService) CreateUser(user domain.User) error {
	fmt.Println("Reached service to create user")

	user.ID = uuid.New().String()
	fmt.Println(user.ID)

	return s.repo.Create(user)

}

func (s *UserService) GetUserById(userID string) (*domain.User, error) {
	fmt.Printf("Fetching user with ID: %s\n", userID)
	return s.repo.GetUserById(userID)
}
func (s *UserService) UpdateUser(user domain.User) error {
	return s.repo.UpdateUser(user)
}

func (s *UserService) DeleteUser(userId string) error {
	return s.repo.Delete(userId)
}

// func (s *UserService) GetUserCountByRole(role string) (int, error) {
// 	return s.repo.CountUsersByRole(role)
// }

func (s *UserService) GetUsersCount() (usercounts domain.UsersCount, err error) {
	return s.repo.GetUsersCount()
}

File: src/services/user_service.py
import uuid
from fastapi import HTTPException, status

from src.models.user import User, UsersCount
from src.errors import error
from src.constants.role_enum import UserRole
from src.utils.mapper import to_user_out


class UserService:
    def __init__(self, user_repo):
        self.user_repo = user_repo

    def get_all_users(self) -> list[User]:
        users = self.user_repo.get_all_users()
        return [to_user_out(u) for u in users if u.Role == UserRole.OWNER]

    def get_user_by_email(self, email: str) -> User:
        user = self.user_repo.get_by_email(email)
        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found",
            )
        return user

    # def create_user(self, user: User) -> None:
    #     user.ID = str(uuid.uuid4())
    #     try:
    #         self.user_repo.create(user)
    #     except error.RepositoryError as e:
    #         raise HTTPException(
    #             status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
    #             detail="Failed to create user",
    #         ) from e

    def get_user_by_id(self, user_id: str) -> User:
        user = self.user_repo.get_user_by_id(user_id)
        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found",
            )
        return to_user_out(user)

    def update_user(self, user: User) -> None:
        self.user_repo.update_user(user)

    def delete_user(self, user_id: str) -> None:
        try:
            self.user_repo.delete(user_id)
        except error.UserNotFoundError:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found",
            )

    def get_users_count(self) -> UsersCount:
        try:
            return self.user_repo.get_users_count()
        except error.RepositoryError as e:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to fetch users count",
            ) from e
